### PR TITLE
Add bidirectional (doubly-linked) list instrusive container template with MIT license.

### PR DESCRIPTION
**Pull request by ywkaras**
I want to add this so @dyrock can use it in this plugin:

https://github.com/dyrock/trafficserver/pull/2

### CODE REVIEW

**Comment by bryancall** [comment]
The way this PR is created it can't be used by a plugin.  Were you going to add it to the C or C++ API?

**Comment by ywkaras** [comment]
@zwoop Are you saying concretely that this PR should be rolled into @dyrock 's certifier PR (that I link to above)?  I thought the rule is that we wanted to make PRs less amalgam-ish rather than more.

**Comment by bryancall** [comment]
I don't see core ever using this internally.

Also, we shouldn't be using ts::string_view or TextView in plugins either.  There is no case that I can see where this is needed.

**Comment by ywkaras** [comment]
Anything that processes HTTP headers is liable to use string_view and TextView if it needs to parse field values.  What should plugins use to parse config files, rather than TextView?

The certifier plugin is using an embedded bidirectional list, rather than std::list, to avoid additional dynamic memory allocation.  This is a fairly common need.

**Comment by ywkaras** [comment]
Here again is the hyperlink to the proposed certifier plugin:  https://github.com/dyrock/trafficserver/pull/2

Which plugins use JSON or YAML for their config files?

**Comment by bryancall** [comment]
2 plugins have json as a dependency and it is a shame every plugin creates it own config language.

I would say that using an intrusive list is preoptimization in the first place.

File: lib/ts/bidir_list.h [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+  */
+
+// Originally copied from https://github.com/wkaras/C-plus-plus-intrusive-container-templates .
+
+/*
+Copyright (c) 2016 Walter William Karas
```
[the remark this answers is not in the document]
Comment (zizhong) [reply lib/ts/bidir_list.h]: right, it would be much helpful if you guys share some good practice to deal with licenses. @zwoop @bryancall

**Comment by ywkaras** [comment]
@bryancall your point about preoptimization seems correct to me, as certification is relatively low-frequency.  @dyrock you should consider replacing the doubly-linked list with std::list<SslData *> .

But when a doubly-linked intrusive list is called for, it's a bad idea to not use a utility for this.  It may not be a lot of code, but the logic is not entirely trivial, easy to make a mistake.  It's not trivial to include automated test cases to check this logic.

**Comment by SolidWallOfCode** [comment]
The bigger problem is what @zwoop pointed out, there is already an intrusive doubly linked list in `lib/ts/List.h`, named `DLL`. An example of its use in a generic way is in `TSHashTable` in `lib/ts/Map.h`.

**Comment by ywkaras** [comment]
Why does List.h have so many #define's in it?

**Comment by ywkaras** [comment]
Why didn't anyone tell @dyrock about List.h ?

**Comment by ywkaras** [comment]
Do the Committers have bets going as to whether they can get @dyrock or me to throw away more code?

**Comment by SolidWallOfCode** [comment]
`List.h` is a bit of a mess with all the preprocessor macros. The goal, as far as I can understand the code, is to make it simpler to have defaults for various names while avoiding name collisions. There's a lot of cleanup that could be done there.

I didn't tell dyrock about List.h because I presumed he'd use an STL container. Doubly linked intrusive lists are a rather specialized type of container and not of much benefit for this particular use.

**Pull request closed without merging**